### PR TITLE
Launch the SPA

### PR DESCRIPTION
**Pull request by kevinshen56714**
This PR sets up the SPA routes, permission checks, and launches the single-page app.

## Routes
1. All the page routes (except for /features, /admin, and /omaha_data) are now handled via a new `SPAHandler` whose template_path is spa.html and doesn't require any page-specific data. The `SPAHandler` is slightly different than the `ConstHandler` in that it has more permission checks (more about permissions below), so I think it's better to separate the two.
2. I separated the GET and POST routes for SPA. The GET routes are added to the Flask app as usual, and the POST routes are provided separately as a list to `basehandlers.FlaskApplication`. These POST routes have the methods=['POST'] and are added ahead of the GET routes so they don't get handled by `SPAHandler`.
3. The POST routes are either handled by the `FeatureCreateHandler` or `FeatureEditHandler` in guide.py (they are just renamed from previous `FeatureNew` and `FeatureEditStage` classes).
4. Some routes were redirected by the server (via `Redirector`), for example, `'/' -> '/roadmap'` and `'/metrics' -> '/metrics/css/popularity'`. These redirections are now handled by the client router in chromedash-app element.

## Permissions
1. In `SPAHandler`, there are three levels of permissions checks - `require_signin`, `require_create_feature`, and `require_edit_feature` that can be specified in the "defaults" dict argument.
2. The /myfeatures and /settings pages have `require_signin` set to True
3. The /guide/new page have `require_create_feature` set to True. The create feature permission is checked by a new method `validate_feature_create_permission` in framework/permissions.py. I had to create this new function because although the current `require_create_feature` function does the same thing, it has to be used as a decorator.
4. The /guide/edit, /guide/editall, /guide/verify_accuracy, and guide/stage pages have `require_edit_feature` set to True. Users need the feature-specific permission to view these pages.
5. Users will be redirected if they don't have the required permissions.

## Misc
1. User object is now fetched through API in `<chromedash-app>`. The user object is now only passed down to the header element, but in a follow-up PR, I will pass it down to all the page elements that fetch user data.
2. `guide.ProcessOverview`, `guide.FeatureEditAllFields`, `guide.FeatureVerifyAccuracy` classes and their unit tests are removed because they don't handle GET request anymore and their POST requests can all be handled the same way via the new `FeatureEditHandler`
3. If this PR looks good. In a follow-up PR, I will remove all the unused HTML files!

### CODE REVIEW

File: static/elements/chromedash-header.js
Comment (jrobbins) [inline static/elements/chromedash-header.js:200]: Is this for MPA pages?  It would be more clear if the comment mentioned "MPA".

**Review by jrobbins — COMMENTED** [review]
This code all looks good.
I'm going to hold off on the LGTM until tomorrow because I want to test it interactively myself, and because I deployed a new version this morning and I want to see if there are any problems that need a quick fix before landing new changes.

**Comment by jrobbins** [comment]
I did manual testing on Chrome, Safari, and Firefox and everything looks good, except one small problem:
I can't star a feature on the feature detail page.  I think the cause is that the feature ID is being passed as a string of digits rather than a JS number. 

Also, this is not a blocker, but there's also some glitches with browser navigation among the metrics pages.  If I start with the stack rank, then drill into a timeline page, I can't get back to the stack rank by pressing the back button once.  I need to hit it twice fast.  And, once I get back, I can't navigate forward via cmd-right-arrow to get back to the timeline.

**Comment by jrobbins** [comment]
I deployed a version to our staging server, so you can see the metrics pages there:
https://d05fbf4-tainted-dot-cr-status-staging.appspot.com/metrics/feature/timeline/popularity/2089

File: main.py
Comment (kevinshen56714) [reply main.py]: Fixed!

File: static/elements/chromedash-header.js
Comment (kevinshen56714) [reply static/elements/chromedash-header.js:200]: Added!

**Comment by kevinshen56714** [comment]
> I did manual testing on Chrome, Safari, and Firefox and everything looks good, except one small problem:
> I can't star a feature on the feature detail page. I think the cause is that the feature ID is being passed as a string of digits rather than a JS number.

Good catch, this is fixed by parseInt the featureId string before passing into the feature detail page component.

> Also, this is not a blocker, but there's also some glitches with browser navigation among the metrics pages. If I start with the stack rank, then drill into a timeline page, I can't get back to the stack rank by pressing the back button once. I need to hit it twice fast. And, once I get back, I can't navigate forward via cmd-right-arrow to get back to the timeline.

This is also fixed. I found that we do history.pushState even when the url already has the correct bucketId, so we end up adding two identical entries to the browser's session history stack. And this happens when a bucketId is provided through url, which is the case when you drill into a timeline page from the stack rank page.

Please review the changes again. Really appreciate it for testing out the SPA!

**Review by jrobbins — APPROVED** [review]
Cue the fireworks!

**Comment by kevinshen56714** [comment]
🎉 🎉 🎉

**Comment by jstenback** [comment]
Exciting! I can't wait to see this live on chromestatus.com!

**Pull request merged**
